### ISSUE TITLE
Intent to support Realm based intrinsics

### ISSUE THREAD

**Issue by jonathanKingston** [babel/babel#868]
As mentioned in: https://github.com/switchfly/ember-cli-mocha/issues/37#issuecomment-75465220 the ECMA6 specification suggests that all intrinsic objects to the browser should be defined based upon their realm and each modules define a new realm.

> From my quick scrambling through the ES6 spec (I'm probably not reading this right).

> [Object Environment Records](http://people.mozilla.org/~jorendorff/es6-draft.html#sec-object-environment-records)
>> Immutable bindings do not exist for object environment records.

> So no help there, however I believe that is just describing normal behaviour of scripts.

> ---

> [Well known intrinsic objects](http://people.mozilla.org/~jorendorff/es6-draft.html#sec-well-known-intrinsic-objects)

>>  Unless otherwise specified each intrinsic object actually corresponds to a set of similar objects, one per Realm.

> [Code Realms](http://people.mozilla.org/~jorendorff/es6-draft.html#sec-code-realms)

>> Before it is evaluated, all ECMAScript code must be associated with a Realm. Conceptually, a realm consists of a set of intrinsic objects, an ECMAScript global environment
> From my quick scrambling through the ES6 spec (I'm probably not reading this right).

> [Object Environment Records](http://people.mozilla.org/~jorendorff/es6-draft.html#sec-object-environment-records)
>> Immutable bindings do not exist for object environment records.

> So no help there, however I believe that is just describing normal behaviour of scripts.

> ---

> [Well known intrinsic objects](http://people.mozilla.org/~jorendorff/es6-draft.html#sec-well-known-intrinsic-objects)

>>  Unless otherwise specified each intrinsic object actually corresponds to a set of similar objects, one per Realm.

> [Code Realms](http://people.mozilla.org/~jorendorff/es6-draft.html#sec-code-realms)

>> Before it is evaluated, all ECMAScript code must be associated with a Realm. Conceptually, a realm consists of a set of intrinsic objects, an ECMAScript global environment

> [Module evaluation](http://people.mozilla.org/~jorendorff/es6-draft.html#sec-moduleevaluation)
>> Set the Realm of moduleCxt to module.[[Realm]].

> ---

>My reading of that is `Object` is considered an **Intrinsic** object which is related to the **Realm** it is in, Modules appear to run in their own **Realm**. Creating a new **Realm** appears to reinstantiate new Intrinsics.

index.js
```
import me from './other';
import me2 from './other2';

console.log('Test1: ' + String.prototype.thing);
```

other.js
```
Object.prototype.thing = 'Shoe';

var test = [];

export default test;
```

other2.js
```
console.log('Test2: ' + String.prototype.thing);
```

```
babel *.js -d out/
```

```
node out/index.js
Test2: Shoe
Test1: Shoe
```

If I am reading the spec correctly, the above code both should return `undefined` not `Shoe`

**Comment by sebmck** [comment]
There's no way to replicate this as it's impossible to have that low level control from user code.

**Comment by jonathanKingston** [comment]
@sebmck if the transpiler ran through the same steps as the engine is supposed to for setting up a new realm it could be possible.

So basically when this feature is enabled, each module redefines all the prototypes back to their default. Any overriding of prototypes would only then impact the module scope.

**Comment by sebmck** [comment]
@jonathanKingston 

> if the transpiler ran through the same steps as the engine is supposed to for setting up a new realm it could be possible.
>
> So basically when this feature is enabled, each module redefines all the prototypes back to their default. 

How?

**Comment by jonathanKingston** [comment]
So like in: [Create intrinsics](http://people.mozilla.org/~jorendorff/es6-draft.html#sec-createintrinsics).

I creating proxy methods to each native type per each module would work.

With code along the lines of this per module:
```
var backup = [];
backup.push(Object.prototype.hasOwnProperty);

//module code
Object.prototype.hasOwnProperty = function () {console.log('sdsd');}

var test = {}
test.hasOwnProperty
//function () {console.log('sdsd');}

Object.prototype.hasOwnProperty = backup[0];
```

**Comment by sebmck** [comment]
That will only work if all the module code is synchronous.

```javascript
Object.prototype.hasOwnProperty = function () { console.log('sdsd'); }

setTimeout(function () {
  {}.hasOwnProperty();
});
```

**Comment by sebmck** [comment]
And in the case of mocha, it's asynchronous.

**Comment by jonathanKingston** [comment]
Hmm yeah, the other idea I had was to redefine the code written itself, so rather than:
```
 var me = {}
```

It would be transpiled into something like:
```
var scopeObject = makeScopeObject();
var me = scopeObject();
```

**Comment by jonathanKingston** [comment]
Which is problematic with included code I am sure.

**Comment by sebmck** [comment]
Monkey patching is a bad idea and there's not going to be a perfect solution to trying to shim this behaviour.

**Comment by jonathanKingston** [comment]
I have fixed this before also using iframe type code which obviously has it's restrictions of use also. Node code has the same with processes too.

Yeah I can see that, I just wanted to raise to see if the idea had been floated, I was happy to be shot down in the process :smile:

**Comment by sebmck** [comment]
Thanks for bringing it up though @jonathanKingston! Always keen to have this type of discussion :stuck_out_tongue_winking_eye:

**Comment by jonathanKingston** [comment]
@sebmck hah yeah I don't even like `chai.should` syntax for this reason, I just wanted to check if my assertion around this being problematic was correct.

Is it worth making clear in the documentation that this isn't supported, to warn code authors when modules are native their code may break?

**Comment by sebmck** [comment]
Potentially, I'm not sure if this is something that warrants a mention though since it's pretty intricate behaviour and it'll likely confuse more users than it'll help.

**Comment by jonathanKingston** [comment]
ember-cli apps would suffer on the move to native because of their prototype methods:
http://emberjs.com/api/classes/Ember.String.html
Also function and array methods which seem to be undocumented.

Is there any other known behaviour that may cause future issues like this?

I'm thinking it might be worthy of a page for library writers as they are likely the ones providing the support for migration to native.

**Comment by sebmck** [comment]
> ember-cli apps would suffer on the move to native because of their prototype methods:
> http://emberjs.com/api/classes/Ember.String.html

That's crazy, had no idea that ember extended native prototypes.

> Also function and array methods which seem to be undocumented.

Such as? Babel doesn't provide anything related to built-ins besides the [polyfill](http://babeljs.io/docs/usage/polyfill) which uses [core-js](https://github.com/zloirock/core-js).

> Is there any other known behaviour that may cause future issues like this?

Not that I know of.

**Comment by jonathanKingston** [comment]
@sebmck no sorry Ember extends array and function prototypes also.

Yeah you can turn it off, which I think might be the correct stance for ember-cli apps given the very justified limitation here.

**Comment by wycats** [comment]
Just to clarify, each ES6 module does not have their own realm. `module.[[Realm]]` is shared across multiple modules, just as multiple scripts loaded for the same page share a realm.

**Comment by jonathanKingston** [comment]
@wycats it does have it's own native though doesn't it? I found it hard to follow in the spec but certainly that is how it read to me.

**Comment by stefanpenner** [comment]
> @wycats it does have it's own native though doesn't it? I found it hard to follow in the spec but certainly that is how it read to me.

Modules have a reference to their realm, but they are not guaranteed their own realm.  For example, one `page`, `frame` or `worker` correlates to 1 realm, all modules contained within a realm, share that realm.

**Comment by stefanpenner** [comment]
> That's crazy, had no idea that ember extended native prototypes.

Ya this is part of our inherited legacy from a time before this was considered a poor idea. Although this is optional, we are planning to eventually remove this pollution.

**Comment by fzzzy** [comment]
What technique does Babel use to support Proxies? The same technique could be used to implement Realm. All the global intrinsic objects would be Proxies instead of plain objects, and the proxy would check the realm and delegate to the actual object for that realm.

Personally I would really love to have multiple cheap global namespaces.

**Comment by sebmck** [comment]
> What technique does Babel use to support Proxies?

None. It would require wrapping every single member expression, assignment etc in helper methods and it would absolutely kill performance so really isn't practical.

**Comment by fzzzy** [comment]
Ah. That's unfortunate.
